### PR TITLE
feat: trial metrics grid (feat @gt2345)

### PR DESCRIPTION
**Pull request by trentwatt**
## Description

Co-authored with Guangqing Tang.

Initial feature-flagged version of trial chart grid.

Does not handle streaming data well: causes an obtrusive flicker and a reset of the zoom. 

I broke the WIP to fix that out into a separate branch, and propose to merge this without that fix since it's behind a feature flag. Ticket is WEB-848.

#5853 addresses some UI issues currently present with the charts.

[WEB-653]
[WEB-660]

image

## Test Plan

To turn on feature flag, append `?f_metrics_grid=on` to the url.

Involves backend changes, so netlify testing is not sufficient.

doing `det e create const.yaml .` from [test-metrics-grid.zip](https://github.com/determined-ai/determined/files/10538838/test-metrics-grid.zip) will generate data that demonstrates the pairing of training/validation metrics.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

[WEB-653]: https://determinedai.atlassian.net/browse/WEB-653?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[WEB-660]: https://determinedai.atlassian.net/browse/WEB-660?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/pages/ExperimentDetails/ExperimentSingleTrialTabs.tsx [hunk 1]
```diff
@@ -77,6 +79,7 @@ const ExperimentSingleTrialTabs: React.FC<Props> = ({
     contextHolder: modalHyperparameterSearchContextHolder,
     modalOpen: openHyperparameterSearchModal,
   } = useModalHyperparameterSearch({ experiment });
+  const chartFlagOn = useFeature().isOn('chart');
```
Comment (gt2345) [inline webui/react/src/pages/ExperimentDetails/ExperimentSingleTrialTabs.tsx:82]: Should this be 'metrics_grid' per PR description?

**Review by gt2345 — APPROVED** [review]
Pairing of training/validation metrics works great!

File: webui/react/src/components/kit/LineChart.tsx [hunk 2]
```diff
@@ -23,9 +23,12 @@ import css from './LineChart.module.scss';
  * @param {string} [name] - Name to display in legend and toolip instead of Series number.
  */
 
+export const TRAIN_PREFIX = /(t_|train_|training_)/;
+export const VAL_PREFIX = /(v_|val_|validation_)/;
```
Comment (mapmeld) [inline webui/react/src/components/kit/LineChart.tsx:27]: if these are here only to import into TrialDetailsOverview, I'd think you might declare them there instead

File: webui/react/src/pages/TrialDetails/useTrialMetrics.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+  const metrics = useMetricNames(trial?.experimentId, handleMetricNamesError);
+  const [data, setData] = useState<Record<MetricName, Serie>>();
+
+  const fetchTrialSummary = useCallback(async () => {
+    if (trial?.id) {
+      const response = await compareTrials({
+        maxDatapoints: screen.width > 1600 ? 1500 : 1000,
+        metricNames: metrics,
+        // scale: scale,
```
Comment (mapmeld) [inline webui/react/src/pages/TrialDetails/useTrialMetrics.ts:69]: intended to comment out?

File: webui/react/src/components/kit/LineChart.tsx (hunk 2 above)
Comment (mapmeld) [reply webui/react/src/components/kit/LineChart.tsx:27]: also would prefer a ^ in the regex, i.e. `/^(t_|train_|training_)/`

File: webui/react/src/components/kit/LineChart.tsx [hunk 4]
```diff
@@ -58,6 +61,8 @@ interface Props {
   yLabel?: string;
 }
 
+export type ChartProps = Props[];
```
Comment (mapmeld) [inline webui/react/src/components/kit/LineChart.tsx:64]: I'm open to this in this file if:
- it's used in `GroupProps` (and related jsdoc) later in this file where it is currently `chartsProps: Props[]`
- the name conveys that it's being used for multiple charts (ChartsProps, ChartGridProps, or MultiChartProps)

File: webui/react/src/pages/TrialDetails/F_TrialDetailsOverview.tsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      const trainingMetricKey = metricToKey(trainingMetric);
+      const trainingMetricSeries = data?.[trainingMetricKey];
+      if (!trainingMetricSeries) return;
+
+      const series: Serie[] = [trainingMetricSeries];
+
+      if (valMetric) {
+        const valMetricKey = metricToKey(valMetric);
+        const valMetricData = data?.[valMetricKey];
+        if (valMetricData) series.push(valMetricData);
```
[the remark this answers is not in the document]
Comment (trentwatt) [reply webui/react/src/pages/TrialDetails/F_TrialDetailsOverview.tsx:60]: In theory, it should be defined. But we are getting the name from one API and the data from another, so seemed safer to have this check.

File: webui/react/src/pages/ExperimentDetails/ExperimentSingleTrialTabs.tsx (hunk 1 above)
Comment (trentwatt) [reply webui/react/src/pages/ExperimentDetails/ExperimentSingleTrialTabs.tsx:82]: I updated as per discussion that there should be only one flag for all the recent chart related changes, see #5876.

File: webui/react/src/pages/TrialDetails/useTrialMetrics.ts (hunk 3 above)
Comment (trentwatt) [reply webui/react/src/pages/TrialDetails/useTrialMetrics.ts:69]: The state for the scale is entirely within the component, so we can't know it from outside. 

In any case, the scale parameter does not affect the correctness of the data being passed back, just the sampling. 

Worth revisiting, some possibilities:
- Make the ChartGrid component a "dumb component" with axes as state that is external and with external handlers (the main pattern in our codebase)
- call API with both values for the scale parameter and put it into a dict (or have the API just return it both way). Not a fan of this option.
- leave as is (maybe non-logarithmic downsampling is sufficient)

Comment (mapmeld) [reply webui/react/src/pages/TrialDetails/useTrialMetrics.ts:69]: I could also see passing some onScaleChange={callback} to the ChartGrid.  But we can make resolving this into its own ticket

**Review by mapmeld — APPROVED** [review]
LGTM, would recc. resolving the one-line merge conflict by accepting the master branch's [x: number, y: number]

File: webui/react/src/pages/TrialDetails/useTrialMetrics.ts (hunk 3 above)
Comment (trentwatt) [reply webui/react/src/pages/TrialDetails/useTrialMetrics.ts:69]: that's what I meant by this:
> axes as state that is external and with external handlers
i think we should hold off till making the scales chart-specific though.

**Review by hamidzr — COMMENTED** [review]
reviewing the master/ changes. Would this double the number of metrics we return? (always both the time-based and batch-based metrics) assuming we want that if we run into perf issues would changing the sampling rate be an option? otherwise it lg

**Comment by trentwatt** [comment]
> reviewing the master/ changes. Would this double the number of metrics we return? (always both the time-based and batch-based metrics) assuming we want that if we run into perf issues would changing the sampling rate be an option? otherwise it lg

yes. the doubling is actually somewhat redundant, since the y values will be the same in both cases, but we want to have all that data client side.

i tested with very high # of steps and it worked fine though. we are looking into unifying/streamlining how we return/consume metrics as a part of Q2 project and revisiting this will be a part of that.

**Comment by ClaireNeveu** [comment]
> the doubling is actually somewhat redundant, since the y values will be the same in both cases
FWIW It's not clear to me how fundamental this redundancy is when accounting for downsampling (is downsampling sensitive to x-axis or just blindly based on number of data points?)

**Comment by trentwatt** [comment]
> > the doubling is actually somewhat redundant, since the y values will be the same in both cases
> > FWIW It's not clear to me how fundamental this redundancy is when accounting for downsampling (is downsampling sensitive to x-axis or just blindly based on number of data points?)

yeah i assume downsampling would be affected, especially if the mapping from batches -> time is highly non-linear, but I dont imagine the end result would be _too_ different

**Comment by trentwatt** [comment]
I don't notice any visual difference when going from downsampling the y values to downsampling their logs, and my assumption would be that the effect there is more pronounced

**Pull request merged**